Elevated access doc (#13046)
dotnet/docs: docs/core/tools/elevated-access.md

**Option 1** Use the full path with elevated prompt:

```cmd
"%ProgramFiles%\dotnet-tools\TOOLCOMMAND"
```

**Option 2** Add the newly created folder to `%Path%`. You only need to do this operation once.

```cmd
setx Path "%Path%;%ProgramFiles%\dotnet-tools\"
```

And run with:

```cmd
TOOLCOMMAND
```

### Uninstall the global tool

In an elevated prompt, type the following command:

```cmd
dotnet tool uninstall PACKAGEID --tool-path "%ProgramFiles%\dotnet-tools"
```

# [Linux](#tab/linux)

[!INCLUDE [elevated-access-unix](../../../includes/elevated-access-unix.md)]

# [macOS](#tab/macos)

[!INCLUDE [elevated-access-unix](../../../includes/elevated-access-unix.md)]

---

## Local tools

Local tools are scoped per subdirectory tree, per user. When run elevated, local tools share a restricted user environment to the elevated environment. In Linux and macOS, this results in files being set with root user-only access. If the user switches back to a restricted account, the user can no longer access or write to the files. So installing tools that require elevation as local tools isn't recommended. Instead, use the `--tool-path` option and the previous guidelines for global tools.

> ## Elevation during development

Author: one more question @wli3 - is this section specific to Linux and macOS?

Reviewer: good question :p it should be able to apply to Windows in principle. I think the current state is good

Author: but isn't sudo something Unix specific?

Reviewer: yes. so on windows, we should open cmd with admin instead. However, we haven't hit this situation for windows IoT yet. So i think we would add it when there is such case (and we can test it)

* If the "Users" group can write or modify the directory, use a different directory name when installing the tools rather than *dotnet-tools*.

To install tools, run the following command in elevated prompt. It will create the *dotnet-tools* folder during the installation.

```cmd
dotnet tool install PACKAGEID --tool-path "%ProgramFiles%\dotnet-tools".
```

### Run the global tool

**Option 1** Use the full path with elevated prompt:

```cmd
"%ProgramFiles%\dotnet-tools\TOOLCOMMAND"
```

**Option 2** Add the newly created folder to `%Path%`. You only need to do this operation once.

```cmd
setx Path "%Path%;%ProgramFiles%\dotnet-tools\"
```

And run with:

```cmd
TOOLCOMMAND
```

### Uninstall the global tool

In an elevated prompt, type the following command:

```cmd
dotnet tool uninstall PACKAGEID --tool-path "%ProgramFiles%\dotnet-tools"
```

# [Linux](#tab/linux)

### Install the global tool

Package assets should be installed in a protected location using the `--tool-path` option. This separation avoids sharing a restricted user environment with an elevated environment.

```bash
sudo dotnet tool install PACKAGEID --tool-path /usr/local/share/dotnet-tools
```

> `/usr/local/share/dotnet-tools` will be created with permission `drwxr-xr-x`. If the directory already exists, use the `ls -l` command to verify the restricted user doesn't have permission to edit the directory. If so, use `sudo chmod o-w -R /usr/share/dotnet-tools` command to remove the access.

Reviewer 2: ```suggestion
`/usr/local/share/dotnet-tools` will be created with permission `drwxr-xr-x`. If the directory already exists, use the `ls -l` command to verify the restricted user doesn't have permission to edit the directory. If so, use the `sudo chmod o-w -R /usr/share/dotnet-tools` command to remove the access.
```

### Run the global tool

> **Option 1** Use full path with sudo:

Reviewer 2: ```suggestion
**Option 1** Use the full path with sudo:
```

```bash
sudo /usr/local/share/dotnet-tools/TOOLCOMMAND
```

> **Option 2** Add symbol link of the tool, once per tool:

Reviewer 2: ```suggestion
**Option 2** Add the symbol link of the tool, once per tool:
```

```bash
sudo ln -s /usr/local/share/dotnet-tools/TOOLCOMMAND /usr/local/bin/TOOLCOMMAND
```

And run with:

```bash
sudo TOOLCOMMAND
```

### Uninstall the global tool

```bash
sudo dotnet tool uninstall PACKAGEID --tool-path /usr/local/share/dotnet-tools
```

If you made a symbol link, you also need to remove it:

```bash
sudo rm /usr/local/bin/TOOLCOMMAND
```

# [macOS](#tab/macos)

### Install the global tool

Package assets should be installed in a protected location using the `--tool-path` option. This separation avoids sharing a restricted user environment with an elevated environment.

```bash
sudo dotnet tool install PACKAGEID --tool-path /usr/local/share/dotnet-tools
```

> `/usr/local/share/dotnet-tools` will be created with permission `drwxr-xr-x`. If the directory already exists, use the `ls -l` command to verify the restricted user doesn't have permission to edit the directory. If so, use `sudo chmod o-w -R /usr/share/dotnet-tools` command to remove the access.

Reviewer 2: ```suggestion
`/usr/local/share/dotnet-tools` will be created with permission `drwxr-xr-x`. If the directory already exists, use the `ls -l` command to verify the restricted user doesn't have permission to edit the directory. If so, use `the sudo chmod o-w -R /usr/share/dotnet-tools` command to remove the access.
```

### Run the global tool

> **Option 1** Use full path with sudo:

Reviewer 2: ```suggestion
**Option 1** Use the full path with sudo:
```

```bash
sudo /usr/local/share/dotnet-tools/TOOLCOMMAND
```

> **Option 2** Add symbol link of the tool, once per tool:

Reviewer 2: ```suggestion
**Option 2** Add the symbol link of the tool, once per tool:
```

```bash
sudo ln -s /usr/local/share/dotnet-tools/TOOLCOMMAND /usr/local/bin/TOOLCOMMAND
```

And run with:

```bash
sudo TOOLCOMMAND
```

### Uninstall the global tool

```bash
sudo dotnet tool uninstall PACKAGEID --tool-path /usr/local/share/dotnet-tools
```

If you made a symbol link, you also need to remove it:

```bash
sudo rm /usr/local/bin/TOOLCOMMAND
```

---

## Local tools

> Local tools are scoped per subdirectory tree, per user. When run elevated, local tools share a restricted user environment to the elevated environment. In Linux and macOS, it results in files being set with root user only access. If the user switch back to a restricted account, the user can no longer access or write to the files. So, installing tools that require elevation as local tools isn't recommended. Instead, use the `--tool-path` option and the previous guidelines for global tools.

Reviewer 2: ```suggestion
Local tools are scoped per subdirectory tree, per user. When run elevated, local tools share a restricted user environment to the elevated environment. In Linux and macOS, this results in files being set with root user-only access. If the user switches back to a restricted account, the user can no longer access or write to the files. So installing tools that require elevation as local tools isn't recommended. Instead, use the `--tool-path` option and the previous guidelines for global tools.
```

## Elevation during development

During development, you may need elevated access to test your application. This scenario is common for IoT apps, for example. We recommend that you build the application without elevation and then run it with elevation. There are a few patterns, as follows:

- Using generated executable (it provides the best startup performance):

> ```bash

Reviewer 2: This needs to be indented.

dotnet build
sudo ./bin/Debug/netcoreapp3.0/APPLICATIONNAME
```

> Using the [dotnet run](dotnet-run.md) command with the `—no-build` flag to avoid generating new binaries:

Reviewer 2: This should be a bulleted list item like the previous one, and the code indented.
